velox/exec/ExchangeClient.cpp

```diff
@@ -196,18 +227,23 @@ ExchangeClient::RequestSpec ExchangeClient::pickSourcesToRequestLocked() {
   RequestSpec toRequest;
   toRequest.maxBytes = averagePageSize;
 
-  // Pick up to 'numToRequest' next sources to request data from.
-  for (auto i = 0; i < sources_.size(); ++i) {
-    auto& source = sources_[nextSourceIndex_];
-
-    nextSourceIndex_ = (nextSourceIndex_ + 1) % sources_.size();
+  // Pick up to 'numToRequest' next sources to request data from. Prioritize
+  // sources that return data.
+  while (toRequest.sources.size() < numToRequest &&
+         !producingSources_.empty()) {
+    auto& source = producingSources_.front();
+    if (source->shouldRequestLocked()) {
+      toRequest.sources.push_back(source);
+    }
+    producingSources_.pop();
+  }
```

Comment:
Should we create a small helper function for this and just call it twice with the different queue object?

Reply:
Let me do that.